tests/integrations/java/test_verify_jdk.py

```diff
@@ -0,0 +1,346 @@
+        returncode=1, cmd='/path/to/java/bin/javac'
+    )
+
+    # Create a directory to make it look like the Briefcase Java already exists.
+    (tmp_path / 'tools' / 'java' / 'bin').mkdir(parents=True)
+
+    # Invoke verify_jdk
+    java_home = verify_jdk(cmd=test_command)
+
+    # The JDK should have the briefcase JAVA_HOME
+    assert java_home == tmp_path / 'tools' / 'java'
+
+    # A single call was made to check javac
+    test_command.subprocess.check_output.assert_called_once_with(
+        [str(Path('/path/to/java/bin/javac')), '-version'],
+        universal_newlines=True,
+        stderr=subprocess.STDOUT,
+    ),
+
+
+def test_unparseable_javac_version(test_command, tmp_path):
+    "If the javac version can't be parsed, the briefcase JDK is used"
+    # Mock getenv returning an explicit JAVA_HOME
+    test_command.os.getenv = mock.MagicMock(return_value='/path/to/java')
+
+    # Mock return value from javac.
+    test_command.subprocess.check_output.return_value = 'NONSENSE\n'
+
+    # Create a directory to make it look like the Briefcase Java already exists.
+    (tmp_path / 'tools' / 'java' / 'bin').mkdir(parents=True)
+
+    # Invoke verify_jdk
+    java_home = verify_jdk(cmd=test_command)
+
+    # The JDK should have the briefcase JAVA_HOME
+    assert java_home == tmp_path / 'tools' / 'java'
+
+    # A single call was made to check javac
+    test_command.subprocess.check_output.assert_called_once_with(
+        [str(Path('/path/to/java/bin/javac')), '-version'],
+        universal_newlines=True,
+        stderr=subprocess.STDOUT,
+    ),
+
+
+@pytest.mark.parametrize(
+    ("host_os, jdk_url, jhome"), [
+        (
+            "Darwin",
+            "https://github.com/AdoptOpenJDK/openjdk8-binaries/releases/download/"
+            "jdk8u242-b08/OpenJDK8U-jdk_x64_mac_hotspot_8u242b08.tar.gz",
+            'java/Contents/Home'
+        ),
+        (
+            "Linux",
+            "https://github.com/AdoptOpenJDK/openjdk8-binaries/releases/download/"
+            "jdk8u242-b08/OpenJDK8U-jdk_x64_linux_hotspot_8u242b08.tar.gz",
+            'java'
+        ),
+        (
+            "Windows",
+            "https://github.com/AdoptOpenJDK/openjdk8-binaries/releases/download/"
+            "jdk8u242-b08/OpenJDK8U-jdk_x64_windows_hotspot_8u242b08.zip",
+            'java'
+        ),
+    ]
+)
+def test_successful_jdk_download(test_command, tmp_path, host_os, jdk_url, jhome):
+    "If needed, a JDK can be downloaded."
+    # Mock host OS
+    test_command.host_os = host_os
+
+    # Mock a JAVA_HOME that won't exist
+    # This is only needed to make macOS *not* run /usr/libexec/java_home
+    test_command.os.getenv = mock.MagicMock(return_value='/does/not/exist')
+
+    # Mock the cached download path
+    archive = mock.MagicMock()
+    archive.__str__.return_value = '/path/to/download.zip'
+    test_command.download_url.return_value = archive
```

Comment:
I agree it's confusing - but it's needed for testing purposes. Otherwise, we have a bunch of places that try to call str(path), and that fails (or, at least, gives unhelpful results for testing purposes)

Reply:
It's not a Path object because we want to validate that unlink/rename is called on the path, and if the path doesn't exist, you can't do that.